### ISSUE TITLE
Wrong gamma conversion formula

### ISSUE THREAD

**Issue by Photosounder** [nothings/stb#588]
Gamma conversion stuff is kind of one of my domains of expertise so I'll tell you about two issues you have with stb_image.h's conversion. Let's look at this line:

`output[i*comp + k] = (float) (pow(data[i*comp+k]/255.0f, stbi__l2h_gamma) * stbi__l2h_scale);`

First of all you're doing it the slow way, for each channel of each pixel you have a `pow()`, which is super slow, a cast and a division. All you need is a 256-entry lookup table of floats with the values precalculated, `data[i*comp+k]` is your index and it gives you directly the converted float value, you can even hardcode the LUT with something like `const float srgb_to_float_table[]={ 0.f, 0.000303527f, 0.000607054f, ... , 0.991102f, 1.f };` (I can generate it for you if you'd like), which saves you the bother of having to worry about concurrency during initialisation.

But much more importantly your formula is wrong, you use a gamma of 2.2 which isn't what we use. Pretty much all computers and image formats for the last 20 years have been using sRGB, which is close to a gamma of 2.2 but isn't a simple gamma, the main difference is in the darks, for instance a value of 1/255 with a gamma of 2.2 would give you a linear value of 0.000005 whereas with the real sRGB formula you'd get 0.0003, which is 60x bigger. I'll just give you the correct formulas in the functions I use:

```c
double lsrgb(double linear)	// converts a [0.0, 1.0] linear value into a [0.0, 1.0] sRGB value
{
	if (linear <= 0.0031308)
		return linear * 12.92;
	else
		return 1.055 * pow(linear, 1.0/2.4) - 0.055;
}

double slrgb(double s)	// converts a [0.0, 1.0] sRGB value into a [0.0, 1.0] linear value
{
	if (s <= 0.04045)
		return s / 12.92;
	else
		return pow((s + 0.055) / 1.055, 2.4);
}
```
So you can use `slrgb((double) i / 255.)` to initialise the aforementioned LUT for the `stbi__ldr_to_hdr` conversion and directly use `lsrgb(data[i*comp+k]) * 255. + 0.5` in `stbi__hdr_to_ldr`.

**Comment by nothings** [comment]
See #508 for why this is complicated and edge-case-y enough that I've never put any time into it.

**Comment by Photosounder** [comment]
Okay, but it's basically a drop-in replacement, you only have two lines to change. But you're right, I guess we should handle each different format with the proper interface anyway.

**Comment by rygorous** [comment]
I'll add that there's another complication, namely that most of the image formats implemented in stb_image predate the sRGB standard, but most of them _do_ have a way to signal gamma as an exponent - and e.g. it used to be that Mac CRTs had a gamma of 1.8 while PC CRTs had around 2.2.

A particular standout example is PNG, which in the 1.0 version of the standard (archived here https://www.w3.org/TR/REC-png-961001) had the gAMA chunk, but no sRGB support. The majority of PNG files produced between 1996 and about 2006 or so had either no gamma information or, if they were sRGB, used a gAMA chunk with exponent of 1/2.2 as the closest possible equivalent. (And then most tools and viewers interpreted that incorrectly, which is another story.)

The short version is that almost every program's color space handling used to be broken in some way or other, and stb_image attempts to sidestep the whole issue by just given you the RGB values stored in the file and leaving all interpretation up to the app.

What's in stb_image is intentionally minimalistic. If you want fast sRGB<->linear conversions (and other more high-level tasks), stb_image_resize has a whole suite, and unlike stb_image they're optimized.

**Comment by Photosounder** [comment]
Good points, I didn't consider the mess that specified gammas create.

**Comment by nothings** [comment]
>The majority of PNG files produced between 1996 and about 2006 or so had either no gamma information or, if they were sRGB, used a gAMA chunk with exponent of 1/2.2 as the closest possible equivalent. (And then most tools and viewers interpreted that incorrectly, which is another story.)

That being said, the original premise of stb_image was "for when you control the images you're loading"--this was to dodge the issue of security, and the fact it only had a few image formats, and to avoid the problem with gamma. Although we do better about all of those now and have reduced the expectation that you authored the images yourself, it is probably not unreasonable for us to support a path where we do parse sRGB or gAMA and adjust the data appropriately, so that a user who *does* have control over all their images and knows they're appropriately tagged can handle them without extra out-of-band data.
